tiledb/sm/query/sparse_index_reader_base.cc

```diff
@@ -181,67 +260,382 @@ Status SparseIndexReaderBase::load_initial_data() {
   }
   RETURN_CANCEL_OR_ERROR(load_tile_offsets(&subarray_, &dim_names_));
 
+  // Compute tile offsets to load and var size to load for attributes.
+  std::vector<std::string> attr_tile_offsets_to_load;
   for (auto& it : buffers_) {
     const auto& name = it.first;
     if (array_schema_->is_dim(name))
       continue;
 
+    attr_tile_offsets_to_load.emplace_back(name);
+
     if (array_schema_->var_size(name))
       var_size_to_load.emplace_back(name);
   }
+
+  // Load tile offsets and var sizes for attributes.
   RETURN_CANCEL_OR_ERROR(load_tile_var_sizes(&subarray_, &var_size_to_load));
+  RETURN_CANCEL_OR_ERROR(
+      load_tile_offsets(&subarray_, &attr_tile_offsets_to_load));
 
   logger_->debug("Initial data loaded");
   initial_data_loaded_ = true;
   return Status::Ok();
 }
 
-// Sort vector elements by second element of tuples.
-bool reverse_tuple_sort_by_second(
-    const tuple<uint64_t, uint64_t, uint64_t>& a,
-    const tuple<uint64_t, uint64_t, uint64_t>& b) {
-  return (std::get<1>(a) > std::get<1>(b));
+Status SparseIndexReaderBase::read_and_unfilter_coords(
+    bool include_coords, const std::vector<ResultTile*>* result_tiles) {
+  auto timer_se = stats_->start_timer("read_and_unfilter_coords");
+
+  // Not including coords or no query condition, exit.
+  if (!include_coords && condition_.empty())
+    return Status::Ok();
+
+  if (subarray_.is_set() || include_coords) {
+    // Read and unfilter zipped coordinate tiles. Note that
+    // this will ignore fragments with a version >= 5.
+    std::vector<std::string> zipped_coords_names = {constants::coords};
+    RETURN_CANCEL_OR_ERROR(
+        read_coordinate_tiles(&zipped_coords_names, result_tiles));
+    RETURN_CANCEL_OR_ERROR(unfilter_tiles(constants::coords, result_tiles));
+
+    // Read and unfilter unzipped coordinate tiles. Note that
+    // this will ignore fragments with a version < 5.
+    RETURN_CANCEL_OR_ERROR(read_coordinate_tiles(&dim_names_, result_tiles));
+    for (const auto& dim_name : dim_names_) {
```

Comment:
Again, no parallelization here. Just making another note on how the task graph will provide a huge parallelization boost soon.

Reply:
https://app.shortcut.com/tiledb-inc/story/12133/sparse-global-order-reader-advanced-optimizations-decide-to-parallelize-on-tiles-dimensions-or-ranges